File: app/src/main/java/jp/gcreate/product/filteredhatebu/model/HatebuBookmark.java
package jp.gcreate.product.filteredhatebu.model;

import com.google.gson.annotations.SerializedName;

/**
 * Copyright 2016 G-CREATE
 */

public class HatebuBookmark {
    public static final HatebuBookmark EMPTY = new HatebuBookmark("", new String[]{""}, "", "");

    @SerializedName("user")
    private String   user;
    @SerializedName("tags")
    private String[] tags;
    @SerializedName("timestamp")
    private String   timestamp;
    @SerializedName("comment")
    private String   comment;

    public HatebuBookmark(String user, String[] tags, String timestamp, String comment) {
        this.user = user;
        this.tags = tags;
        this.timestamp = timestamp;
        this.comment = comment;
    }

    public String getUser() {
        return user;
    }

    public void setUser(String user) {
        this.user = user;
    }

    public String[] getTags() {
        return tags;
    }

    public void setTags(String[] tags) {
        this.tags = tags;
    }

    public String getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(String timestamp) {
        this.timestamp = timestamp;
    }

    public String getComment() {
        return comment;
    }

    public void setComment(String comment) {
        this.comment = comment;
    }

    @Override
    public String toString() {
        return "(" + getClass().getSimpleName() + ":" + hashCode() + "){"
               + "user:" + user + ","
               + "tags:" + tags + ","
               + "timestamp:" + timestamp + ","
               + "comment:" + comment + "}";
    }
}

File: app/src/main/java/jp/gcreate/product/filteredhatebu/model/HatebuBookmark.kt
package jp.gcreate.product.filteredhatebu.model

import com.squareup.moshi.JsonClass

/**
 * Copyright 2016 G-CREATE
 */
@JsonClass(generateAdapter = true)
data class HatebuBookmark(
    var user: String,
    var tags: Array<String>,
    var timestamp: String,
    var comment: String
) {
    companion object {
        val EMPTY = HatebuBookmark("", arrayOf(""), "", "")
    }
    
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        
        other as HatebuBookmark
        
        if (user != other.user) return false
        if (timestamp != other.timestamp) return false
        if (comment != other.comment) return false
        
        return true
    }
    
    override fun hashCode(): Int {
        var result = user.hashCode()
        result = 31 * result + timestamp.hashCode()
        result = 31 * result + comment.hashCode()
        return result
    }
}

sealed class HatebuComments {
    object Disallow: HatebuComments()
    
    data class Comments(val comments: List<HatebuBookmark>): HatebuComments()
    
    object Empty : HatebuComments()
    
    data class Error(val cause: Exception): HatebuComments()
    
    object Loading : HatebuComments()
}
